nep-20.mediawiki

```diff
@@ -0,0 +1,533 @@
+
+While there may be several options for 3rd party wallet providers that will help them to facilitate these transactions, there is currently no common consensus on the consistency of their interfaces. With a lack of consistency in interfaces to use these wallet provider services, dApp developers will be forced to make a decision to have their platform supported by a single provider, or to double or even triple their development efforts to accommodate all the different wallet provider interfaces. This will lead to a fragmentation in the dApps ecosystem.
+
+===Wallet providers===
+
+Each wallet provider, when deciding on supporting dApps to utilize their services as an authentication mechanism will be faced with a decision on how to implement an API to communicate with the dApps. Wallet providers can choose to create their own API from scratch, create their own version of existing projects, or aim to directly duplicate an existing API. In the case that the provider decides to make their own API interface from scratch, and try to promote dApps to use it, time and effort will inevitably be wasted by both the provider and competing providers on getting dApp developers on board with using their custom communication interface. If we have a unified interface for such transactions, providers could spend more time on making their individual services better for their users.
+
+==Specification==
+
+===Types===
+
+<pre>
+// Represents a piece of data encoded as a base64 string.
+export type Base64Encoded = string;
+
+// Represents a N3 address.
+// Example: "NSiVJYZej4XsxG5CUpdwn7VRQk8iiiDMPM"
+export type Address = string;
+
+// A 160-bit hash represented by a hexadecimal string.
+// Example: "0x682cca3ebdc66210e5847d7f8115846586079d4a"
+export type UInt160 = string;
+
+// A 256-bit hash represented by a hexadecimal string.
+// Example: "0x1f4d1defa46faa5e7b9b8d3f79a06bec777d7c26c4aa5f6f5899a291daa87c15"
+export type UInt256 = string;
+
+// Represents an ECC public key.
+// Example: "03b209fd4f53a7170ea4444e0cb0a6bb6a53c2bd016926989cf85f9b0fba17a70c"
+export type ECPoint = string;
+
+// A large integer of any length expressed as a string.
+export type Integer = string;
+
+// Represents a piece of data encoded as a hexadecimal string.
+export type HexString = string;
+
+// Represents a version.
+// Example: "1.0.0"
+export type Version = string;
+
+// Represents the N3 network.
+// MAINNET: 860833102
+// TESTNET: 877933390
+export type Network = number;
+
+// Represents the name of the event in IDapiProvider.
+export type EventName =
+    "accountschanged" | // Triggered when the accounts in the wallet change.
+    "networkchanged";   // Triggered when the user switches the current network.

```

Comment:
Could we have a new event for `connectionProposal` that could be emitted to deliver a connection code? The dapp could show it as QRCode to stablish a connection. This is very important for mobile Wallets.

Reply:
I don't quite understand the role of `connectionProposal`. Can you draw a flowchart to explain it?

Reply:
- Hey, I just made this, I hope it could be helpful:

![wallet-connect-sequence-diagram](https://user-images.githubusercontent.com/1337983/145034246-58b4b9d5-237a-4be7-bcf8-9f6b6eb93f3a.jpg)

- Also, I wrote about it on [this Readme](https://github.com/CityOfZion/wallet-connect-sdk)

---

- And also, I wrote this while I was investingating it:

# Wallet Connect implementation
Wallet Connect is a protocol to allow a safer communication between dApp and Wallet. It works with any blockchain because it simply creates a WebSocket server that handles the communication between dApp and Wallet in a smart way, handling format validation, problems with internet connection, cache, log, etc. It's client-side libraries helps to integrate even further, allowing the user to scan a QR Code to connect the two ends.

## The parts
![](https://gblobscdn.gitbook.com/assets%2F-LJJeCjcLrr53DcT1Ml7%2Fsync%2F95c321d277f8179985ee573e3d7f7d7074c319b4.png?alt=media)

### Server
A simple middleman that handles the communication between the other parts. At first, we believe we would need to change Wallet Connect relay server so it could work with Neo Blockchain and then send a pull request. But after exploring its code, it looks like we can simply use Wallet Connect server as it is right now.

### dApp
A dApp is an application that uses smart contracts and other technologies as server-side. To securely reach the Smart Contracts, the user must use its own Wallet to sign the operations, and the responsible to delegate such interaction is the dApp's client-side. Firstly, the dApp establishes a connection with Wallet Connect which asks the user to scan the QR Code, by doing this, Wallet Connect knows how to make an end-to-end encryption and transfer information between the dApp and the Wallet. Now, the dApp can request to do an operation on the blockchain using the user's Wallet. By using JsonRPC format, it says which method and arguments it wants to call.

### Wallet
The Wallet is a client-side application that allows the user to handle its assets, sign documents and interact with Smart Contracts. If integrated with Wallet Connect, it can receive messages from dApps that requests the user to procced with a blockchain operation. When receiving such methods, the Wallet should ask for user authorization and then call the blockchain method by itself. A good UX pattern is to show a mobile notification so the user can allow it without having to open the app by himself.

Reply:
As I said:

> Dapi is a protocol for communication between plug-in wallet and dapp via javascript. And WalletConnect allows communication between the two through a centralized server to relay messages.

They are different.

Reply:
Sure, but they focus solving the same problem. Are we going to exclude the possibility for WalletConnect integration on this specification? I mean, if the dapp wants to have multiple dapp-wallet integration providers, WalletConnect couldn't be one of them? Meaning that there will be absolutely no mobile wallet providers.